src/image/channel_groups.rs

```diff
@@ -155,43 +240,26 @@ impl<'slf, ChannelGroup> WritableChannels<'slf> for ChannelGroups<ChannelGroup>
         mode
     }
 
-    type Writer = GroupChannelsWriter<'slf, ChannelGroup>;
+    type Writer = GroupChannelsWriter<ChannelGroup::Writer>;
 
     fn create_writer(&'slf self, header: &Header) -> Self::Writer {
-        let channels = header.channels.list.iter()
-            .map(|channel_info|{
-                // hashmap order is not guaranteed? so look up each channel group manually instead of generating new
-                let channels = self.lookup_group(channel_info.name.as_slice())
-                    .expect("channels not found bug");
-
-                channels.create_writer(header) // channel_info.name.clone()
-            })
-            .collect();
-
-        GroupChannelsWriter { channels_list: channels }
+        GroupChannelsWriter {
+            all_channel_groups: self.all_channel_groups()
+                .map(|channel_group: &ChannelGroup| channel_group.create_writer(header))
+                .collect()
+        }
     }
 }
 
-struct GroupChannelsWriter<'c, ChannelGroupWriter> {
-    channels_list: Vec<&'c ChannelGroupWriter>,
+pub struct GroupChannelsWriter<ChannelGroupWriter> {
+    all_channel_groups: Vec<ChannelGroupWriter>,
 }
 
-impl<'c, Channels> ChannelsWriter for GroupChannelsWriter<'c, Channels> where Channels: ChannelsWriter {
-    fn extract_uncompressed_block(&self, header: &Header, block: BlockIndex) -> Vec<u8> {
-        let mut blocks_per_channel: Vec<Cursor<Vec<u8>>> = self
-            .channels_list.iter()
-            .map(|channels| Cursor::new(channels.extract_uncompressed_block(header, block)))
-            .collect();
-
-        UncompressedBlock::uncompressed_block_from_lines(header, block, |line|{
-            let channel_reader = &mut blocks_per_channel[line.location.channel]; // TODO subsampling
-
-            // read from specific channel into total byte block
-            // this assumes that the lines in the callback are iterated in strictly increasing order
-            // because each channel reader is consumed
-            channel_reader.read_exact(line.value)
-                .expect("collecting grouped channel byte block failed");
-        })
+impl<Channels> ChannelsWriter for GroupChannelsWriter<Channels> where Channels: ChannelsWriter {
+    fn extract_uncompressed_block(&self, header: &Header, block_index: BlockIndex, output_block_data: &mut [u8]) {
+        for channels_group in &self.all_channel_groups {
```

Comment:
Each subgroup will write to the same few first bytes, overwriting each other. Incorporate some kind of offset?